app/src/main/java/org/oppia/app/parser/StringToNumberWithUnitsParser.kt

```diff
@@ -0,0 +1,168 @@
+    if (rawInput.isNotEmpty() && rawInput != null) {
+      // Start with digit when there is no currency unit.
+      if (rawInput.matches("^\\-?\\d.*\$".toRegex())) {
+        var ind = indexOf(Pattern.compile("[a-z(₹$]"), rawInput)
+        if (ind == -1) {
+          // There is value with no units.
+          value = rawInput;
+          units = ""
+        } else {
+          ind -= 1
+          value = rawInput.substring(0, ind).trim()
+          units = rawInput.substring(ind).trim()
+        }
+
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (units.indexOf(CURRENCY_UNITS[i]!!.frontUnits[j]) != -1) {
+              throw  Error(
+                "INVALID_CURRENCY_FORMAT"
+              )
+            }
+          }
+        }
+      } else {
+        var startsWithCorrectCurrencyUnit = false
+        var keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (rawInput.startsWith(CURRENCY_UNITS[i]!!.frontUnits[j])) {
+              startsWithCorrectCurrencyUnit = true
+              break
+            }
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+
+        var ind = indexOf(Pattern.compile("[0-9]"), rawInput)
+
+
+        if (ind === -1) {
+          throw  Error("INVALID_CURRENCY")
+        }
+        units = rawInput.substring(0, ind).trim()
+
+        startsWithCorrectCurrencyUnit = false
+        keys = (CURRENCY_UNITS).keys
+        for (i in keys) {
+          for (j in 0 until CURRENCY_UNITS[i]!!.frontUnits.size) {
+            if (units == (CURRENCY_UNITS[i]!!.frontUnits[j]).trim()) {
+              startsWithCorrectCurrencyUnit = true
+              break
+            }
+          }
+        }
+        if (startsWithCorrectCurrencyUnit == false) {
+          throw  Error("INVALID_CURRENCY")
+        }
+        units = units + ""
+
+        var ind2 = indexOf(Pattern.compile("[a-z(]"), rawInput.substring(ind))
+        if (ind2 != -1) {
+          value = rawInput.substring(ind, ind2 - ind).trim()
+          units += rawInput.substring(ind2).trim()
+        } else {
+          value = rawInput.substring(ind).trim()
+          units = units.trim()
+        }
+      }
+      // Checking invalid characters in value.
+      if (value.matches("[a - z]".toRegex()) || value.matches("[ *^$₹()#@]/".toRegex())) {
+        throw  Error("INVALID_VALUE")
+      }
+
+      if (value.contains('/')) {
+        type = "fraction"
+        fractionObj = StringToFractionParser().parseFractionFromString(value)
+      } else {
+        type = "real"
+        real = Double.parseDouble(value)
+      }
+      if (units != "") {
+        // Checking invalid characters in units.
+        if (units.matches("[^0-9a-z/* ^()₹$-]".toRegex())) {
+          throw Error(
```

Comment:
This error should be removed to be real-time or submit-time error. Ditto for all other errors above.

Reply:
@BenHenning I need to work on final validation and real-time or submit-time error messages once the parsing approach is finalised as communicated already... is that ok?